### PR TITLE
FloatingPoint: Separate out symFPU glue code.

### PR DESCRIPTION
**Pull request by aniemetz**
This works towards having the symFPU headers only included where it's
absolutely needed (only in the .cpp files, not in any other headers).

Note: src/util/floatingpoint.h.in will be converted to a regular .h file
      in the next step to simplify reviewing.

### CODE REVIEW

File: src/util/floatingpoint.cpp [hunk 1]
```diff
@@ -64,203 +69,232 @@ CVC4_LIT_ITE_DFN(::CVC4::symfpuLiteral::traits::ubv);
...
-      for (uint32_t i = 0; i < sigBits - 1; ++i)
+    // Work out the number of bits required to represent the exponent for a
+    // normal number
+    uint32_t expBits = 2;  // No point starting with an invalid amount
+
+    Integer doubleInt(2);
+    if (exp.strictlyPositive())
+    {
+      Integer representable(
+          4);  // 1 more than exactly representable with expBits
```
Comment (martin-cs) [inline src/util/floatingpoint.cpp:215]: Kind of funky formatting.

**Review by martin-cs — APPROVED** [review]
This is great.  I tried to get SymFPU out of the headers but only managed in the symbolic case.  This does the literal case as well which is brilliant.  Thanks!

File: src/util/floatingpoint.cpp [hunk 2]
```diff
@@ -64,203 +69,232 @@ CVC4_LIT_ITE_DFN(::CVC4::symfpuLiteral::traits::ubv);
...
-	  working /= two;
-	}
-      } else {
-	while (r >= working) {
-	  exp += inc;
-	  working *= two;
-	}
-	exp -= inc;
-	working /= two;
+    if (rabs != working)
```
[the remark this answers is not in the document]
Comment (aniemetz) [reply src/util/floatingpoint.cpp:182]: I made 2 changes:
You had
```
if (r == working)
{
}
else if ()
{
}
else
{
}
```
and I wrapped this in `if (r != working) {...}`.

The other one was that you had `Rational r(ri)` (where `ri` was the one you passed in in the constructorHelper function).
This only changes it to be `r = r.abs()` (where r is overwritten and that's why you had to copy it.

I'm actually in the process of starting debug cluster runs right now, wanted to get an opinion from you before you go to bed, hence already submitted the PR :-)

[the remark this answers is not in the document]
Comment (martin-cs) [reply src/util/floatingpoint.cpp:192]: Yeah; this should be fine.

[the remark this answers is not in the document]
Comment (aniemetz) [reply src/util/floatingpoint.cpp:192]: Ran it on non-incremental QF_FP and QF_BVFP in debug, no errors.

[the remark this answers is not in the document]
Comment (martin-cs) [reply src/util/floatingpoint.cpp:192]: This is good but that code we are discussing shouldn't be run / needed for QF_FP and QF_BVFP.  You need *FP*RA* or similar.

[the remark this answers is not in the document]
Comment (aniemetz) [reply src/util/floatingpoint.cpp:192]: Shoot, you're right. My brain is already in weekend mode. Started.

[the remark this answers is not in the document]
Comment (aniemetz) [reply src/util/floatingpoint.cpp:192]: No errors on non-incremental QF_FPLRA and QF_BVFPLRA.

[the remark this answers is not in the document]
Comment (martin-cs) [reply src/util/floatingpoint.cpp:192]: The key one is Flo's test suite for `to_fp` from `Real`.  If all of that passes you should be good to go.

**Pull request merged**